flink-core/src/main/java/org/apache/flink/api/common/typeutils/CompositeTypeSerializerSnapshot.java

```diff
@@ -155,7 +155,7 @@ public final void readSnapshot(int readVersion, DataInputView in, ClassLoader us
 	TypeSerializerSchemaCompatibility<T> internalResolveSchemaCompatibility(
 			TypeSerializer<T> newSerializer,
 			TypeSerializerSnapshot<?>[] snapshots) {
-		if (newSerializer.getClass() != correspondingSerializerClass) {
+		if (!correspondingSerializerClass.isInstance(newSerializer)) {
```

Comment:
It is here because some of the tuple serializers snapshot are using an instance of  `SpecificCaseClassSerializer` (i.e. `Tuple2CaseClassSerializerSnapshot`) during the migration path from pre 1.8 (where previously all the tuple related serializers used the same `TypeSerializerConfigSnapshot`)

Reply:
I don't fully understand why this change is required.

for the `Tuple2CaseClassSerializerSnapshot`, the `correspondingSerializerClass` is a `Tuple2CaseClassSerializer`, and so is the new serializer, no?

Moreover, this breaks the previous compatibility contract of the `CompositeTypeSerializerSnapshot` - serializers that return this snapshot are not expected to change.

Reply:
I think that I've provide to Aljoscha the wrong example, the correct example would be the anonymous case class serializer that returns `SpecificCaseClassSerializerSnapshot` in `snapshotConfiguration`, which leads to having a snapshot class that is not compatible with the serializer (without the change) 
But might be this is not a real issue since the checks goes in the other direction (restored snapshot class vs new serializer)
If you think it is safe to drop it, then feel free to exclude this commit :-)

Reply:
Yes, the reason I think this isn't really required is because the new concrete `SpecificCaseClassSerializer` should always be the one that is provided to the check. Therefore, we would never have subclasses here.

I'll remove it and run some tests to be sure. Thanks.